File: TrnAPI_01/BLL/Services/Implementations/RecordService.cs
using BLL.Services.Interfaces;
using Common.Model;
using DAL.Repositories.Interfaces;
using System.Collections.Generic;

namespace BLL.Services.Implementations
{
    public class RecordService : IRecordService
    {
        private readonly IRecordRepository repository;

        public RecordService(IRecordRepository repository)
        {
            this.repository = repository;
        }

        public void AddEvent(int recordId, int eventId)
        {
            repository.AddEvent(recordId, eventId);
        }

        public void RemoveEvent(int recordId, int eventId)
        {
            repository.RemoveEvent(recordId, eventId);
        }

        public void Create(Record item)
        {
            repository.Create(item);
        }

        public void Delete(int id)
        {
            repository.Delete(id);
        }

        public IEnumerable<Record> GetAll()
        {
            return repository.GetAll();
        }

        public Record GetById(int id)
        {
            return repository.GetById(id);
        }

        public IEnumerable<Record> GetRecordsByCategoryId(int categoryId)
        {
            return repository.GetRecordsByCategoryId(categoryId);
        }

        public void Update(Record item)
        {
            repository.Update(item);
        }
    }
}

File: recordbook/src/app/services/recordService.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Record } from '../model/record';
import { RecordViewModel } from '../model/recordViewModel';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { SimpleService } from './simpleService';
import { catchError } from 'rxjs/operators';
import { EventModel } from '../model/EventModel';
import { CategoryService } from './categoryService';
import { Category } from '../model/category';
import { EventService } from './eventService';

@Injectable({
    providedIn: 'root'
})
export class RecordService extends SimpleService<Record> {
    private urlSuffics = 'record';

    constructor(private http: HttpClient) {
        super(http);
    }

    public getAllRecords(): Observable<Record[]> {
        return super.getAll(this.urlSuffics);
    }

    public getRecordById(id: number): Observable<Record> {
        return super.getById(this.urlSuffics, id);
    }

    public updateRecord(model: Record): Observable<any> {
        return super.update(this.urlSuffics, model, model.Id);
    }

    public addRecord(model: Record): Observable<Record> {
        return super.add(this.urlSuffics, model);
    }

    public deleteRecord(id: number): Observable<Record> {
        return super.delete(this.urlSuffics, id);
    }

    public addEvent(recordId: number, eventId: number): Observable<EventModel> {
        const url = `${this.baseUrl}/${this.urlSuffics}/addevent/${recordId},${eventId}`;
        return this.http.post<EventModel>(url, null).pipe(catchError(this.handleError<EventModel>(`addEventToRecord recordId=${recordId}`)));
    }

    public removeEvent(recordId: number, eventId: number): Observable<EventModel> {
        const url = `${this.baseUrl}/${this.urlSuffics}/removeevent/${recordId},${eventId}`;
        const httpOptions = {
            headers: new HttpHeaders({ 'Content-Type': 'application/json' })
        };
        return this.http.delete<EventModel>(url, httpOptions)
            .pipe(catchError(this.handleError<EventModel>(`addEventToRecord recordId=${recordId}`)));
    }

    public getAllRecordViewModels(): RecordViewModel[] {
        let records: Record[];
        let result: RecordViewModel[];
        super.getAll(this.urlSuffics).subscribe((data: Record[]) => records = data);

        let categoryService = new CategoryService(this.http);
        let eventService = new EventService(this.http);

        records.forEach(function (record) {
            let recordViewModel = new RecordViewModel();
            recordViewModel.Id = record.Id;
            recordViewModel.Name = record.Name;
            recordViewModel.CreateDate = record.CreateDate;
            recordViewModel.Description = record.Description;

            categoryService.getCategoryById(record.CategoryId)
                .subscribe((data: Category) => recordViewModel.Category = data.Name);

            let events: EventModel[];
            eventService.getEventByRecordId(record.Id)
                .subscribe((data: EventModel[]) => recordViewModel.Events = data);

            result.push(recordViewModel);
        });

        return result;
    }
}
